[IDE] Set solution-specific variable types as interface types

ahoppen opened the pull request:
Setting the interface type of a variable, just to reset it to a null type is actually really gross. But quite a few methods further down in the generation of code completion results (such as USR generation) need to get a variable’s type and passing them along in a separate map would be really invasive. So this seems like the least bad solution to me.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

rintaro commented on include/swift/IDE/TypeCheckCompletionCallback.h:
```diff
@@ -80,7 +80,31 @@ Type getPatternMatchType(const constraints::Solution &S, Expr *E);
+  }
+
+  ~WithSolutionSpecificVarTypesRAII() {
+    for (auto Var : VarsWithSolutionSpecificTypes) {
```
```suggestion
    for (auto *Var : VarsWithSolutionSpecificTypes) {
```

rintaro replied:
```suggestion
    for (auto *Var : VarsWithSolutionSpecificTypes) {
```
to clarify this is a pointer.

rintaro commented on include/swift/IDE/TypeCheckCompletionCallback.h:
```diff
@@ -80,7 +80,31 @@ Type getPatternMatchType(const constraints::Solution &S, Expr *E);
+
+  ~WithSolutionSpecificVarTypesRAII() {
+    for (auto Var : VarsWithSolutionSpecificTypes) {
+      Var->setInterfaceType(Type());
```
Don't we need to reset this to the pre-RAII value? Or they are guaranteed to be null types? If so, could you add assertion in the constructor?

ahoppen commented:
Note to self: The stress tester passed for this PR

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on include/swift/IDE/TypeCheckCompletionCallback.h:
```diff
@@ -78,9 +78,44 @@ Type getPatternMatchType(const constraints::Solution &S, Expr *E);
+void getSolutionSpecificVarTypes(const constraints::Solution &S,
+                                 llvm::SmallDenseMap<VarDecl *, Type> &Result);
+
+/// While this RAII is alive the interface types of the variables defined in
+/// \c SolutionSpecificVarTypes are reset to null types.
+/// Afterwards, their types are reset.
```
```suggestion
/// While this RAII is alive the interface types of the variables defined in
/// \c SolutionSpecificVarTypes are temporality set to the type in the map.
/// Afterwards, their types are restored.
```

rintaro commented on include/swift/IDE/ArgumentCompletion.h:
```diff
@@ -52,7 +52,7 @@ class ArgumentTypeCheckCompletionCallback : public TypeCheckCompletionCallback {
     /// this result. This in particular includes parameters of closures that
     /// were type-checked with the code completion expression.
-    llvm::SmallDenseMap<const VarDecl *, Type> SolutionSpecificVarTypes;
+    llvm::SmallDenseMap<VarDecl *, Type> SolutionSpecificVarTypes;
```
Non-`const` only for `InterfaceTypeRequest{VD}` in `WithSolutionSpecificVarTypesRAII::setInterfaceType`? If so, I would `const_cast` there, just like `ValueDecl::getInterfaceType()` ;)

ahoppen committed:
[IDE] Set solution-specific variable types as interface types

Setting the interface type of a variable, just to reset it to a null type is actually really gross. But quite a few methods further down in the generation of code completion results (such as USR generation) need to get a variable’s type and passing them along in a separate map would be really invasive. So this seems like the least bad solution to me.

ahoppen commented:
@swift-ci Please smoke test
